src/runner/browser-set.js

```diff
@@ -0,0 +1,130 @@
+import { EventEmitter } from 'events';
+import { Promise } from 'es6-promise';
+import once from 'once';
+import LocalBrowserConnection from '../browser-connection/local';
+import { MESSAGE, getText } from '../messages';
+import remove from '../utils/array-remove';
+
+
+export default class BrowserSet extends EventEmitter {
+    constructor (connections) {
+        super();
+
+        this.BROWSER_CONNECTION_READY_TIMEOUT = 30000;
+        this.WAITING_FOR_DISPOSE_TIMEOUT      = 10000;
+
+        this.promisedDisposals = [];
+
+        this.connections = connections;
+
+        this.browserErrorHandler = msg => this.emit('error', msg);
+
+        connections.forEach(bc => bc.on('error', this.browserErrorHandler));
+
+        //NOTE: We're setting an empty error handler, because Node kills the process on an 'error' event
+        //if there is no handler. See https://nodejs.org/api/events.html#events_class_events_eventemitter
+        this.on('error', () => {});
+    }
+
+    _waitIdle (bc) {
+        return new Promise(resolve => {
+            if (bc.idle || bc.closed || !bc.ready) {
+                resolve();
+                return;
+            }
+
+            var eventHandler = once(() => resolve());
+
+            bc.once('idle', eventHandler);
+            bc.once('closed', eventHandler);
+        });
+    }
+
+    _closeConnection (bc) {
+        return new Promise(resolve => {
+            if (bc.closed || !bc.ready) {
+                resolve();
+                return;
+            }
+
+            bc.close();
+            bc.once('closed', resolve);
+        });
+    }
+
+    _waitConnectionsReady () {
+        return Promise.race([
+            Promise.all(
```

Comment:
If we'll have variables for both promises with the meaningful names, then this will look really handy.